### PR TITLE
4q state prep

### PR DESCRIPTION
**Pull request by kilimanjaro**
This implements the 4q state prep method mentioned in https://github.com/rigetti/quilc/issues/76.
It also updates the test that https://github.com/rigetti/magicl/pull/58 breaks.

### CODE REVIEW

**Comment by notmgsk** [comment]
Heads up: think you will need to release a new magicl and bump the dependency here.

File: src/compilers/state-prep.lisp [hunk 1]
```diff
@@ -232,6 +232,92 @@
...
                                      :arguments (list (first (application-arguments instr))))))))))
 
+(defun coerce-to-complex-double-vector (elts)
+  (map '(vector (complex double-float))
+       (lambda (val) (coerce val '(complex double-float)))
+       elts))
+
+(defun schmidt-decomposition (phi num-a num-b)
+  "Given a wavefunction PHI containing subystems of size NUM-A and NUM-B qubits, compute the Schmidt decomposition of PHI.
+Returns c, U, V, where c is vector and U,V are unitary matrices (of dimensions
```
Comment (notmgsk) [inline src/compilers/state-prep.lisp]: "... is *a* vector ...", space in "U,V"

don't break lines

File: src/compilers/state-prep.lisp [hunk 2]
```diff
@@ -232,6 +232,92 @@
...
+   PHI = \sum_{i} d_i U_i V _i
+
+where U_i, V_i denotes the ith column of the matrix U, V respectively."
+  (assert (= (qubit-count phi) (+ num-a num-b)))
+  (when (listp phi)
+    (setf phi (coerce-to-complex-double-vector phi)))
+  (let* ((size-a (expt 2 num-a))
+         (size-b (expt 2 num-b))
+         ;; tranpose here to make this row-major
+         (reshaped (magicl:transpose (magicl:make-matrix :rows size-a
```
Comment (notmgsk) [inline src/compilers/state-prep.lisp]: why not a complex vector here?

File: src/compilers/state-prep.lisp [hunk 3]
```diff
@@ -232,6 +232,92 @@
...
+        (magicl:svd reshaped)
+      (values (coerce-to-complex-double-vector (magicl:matrix-diagonal d))
+              u
+              (magicl:transpose vt)))))
+
+(defun state-prep-4q (wf q0 q1 q2 q3 &key reversed)
+  "Produce instructions to prepare the wavefunction WF with respect to qubits Q0
+Q1 Q2 Q3, starting from the zero state. If REVERSED is T, the instructions
+instead prepare the zero state from WF."
+   ;;; The following is from arXiv:1003.5760
```
Comment (notmgsk) [inline src/compilers/state-prep.lisp]: weird indent. Two colons

File: src/compilers/state-prep.lisp [hunk 4]
```diff
@@ -232,6 +232,92 @@
...
+  ;; flag on its own
+  (flet ((state-prep-2q (source target qubit-indices)
+           (when reversed
+             (rotatef source target))
+           (make-instance 'state-prep-application
+                          :source-wf source
+                          :target-wf target
+                          :arguments (mapcar #'qubit qubit-indices)))
+         (cnot (a b)
+           (build-gate "CNOT" nil a b))
```
Comment (notmgsk) [inline src/compilers/state-prep.lisp]: stylistically `nil` -> `()`

File: src/matrix-operations.lisp [hunk 5]
```diff
@@ -182,6 +182,21 @@ as needed so that they are the same size."
...
+  (let* ((size (expt 2 n-qubits)))
+    (loop :repeat size
+          :for c := (complex (alexandria:gaussian-random)
+                                 (alexandria:gaussian-random))
+          :collect c :into entries
+          :sum (expt (abs c) 2) :into norm-squared
+          :finally
+             (return (make-array size
+                                 :initial-contents
+                                 (let ((scaling-factor (/ 1 (sqrt norm-squared))))
```
Comment (notmgsk) [inline src/matrix-operations.lisp]: `(/ 1 c)` -> `(/ c)`

File: tests/state-prep-tests.lisp [hunk 6]
```diff
@@ -4,23 +4,34 @@
 
 (in-package #:cl-quil-tests)
 
+(defun wf-to-matrix (wf)
+  "Convert a sequence WF to a corresponding column vector."
+  (magicl:make-matrix :rows (length wf)
```
Comment (notmgsk) [inline tests/state-prep-tests.lisp]: complex?

File: tests/state-prep-tests.lisp [hunk 7]
```diff
@@ -4,23 +4,34 @@
...
+                      :cols 1
+                      :data (copy-seq wf)))
+
+(defun check-state-prep (source-wf target-wf matrix)
+  "Checks whether SOURCE-WF maps to TARGET-WF under the specified MATRIX."
+  (let* ((result (magicl:multiply-complex-matrices matrix
```
Comment (notmgsk) [inline tests/state-prep-tests.lisp]: `(quil:m* matrix-a matrix-b ...)`

File: src/matrix-operations.lisp [hunk 8]
```diff
@@ -182,6 +182,21 @@ as needed so that they are the same size."
   (let ((m (magicl:random-unitary n)))
     (magicl:scale (expt (magicl:det m) (/ (- n))) m)))
 
+(defun random-wavefunction (n-qubits)
+  "Get a random complex unit vector with (EXPT 2 N-QUBITS) entries."
+  (let* ((size (expt 2 n-qubits)))
+    (loop :repeat size
+          :for c := (complex (alexandria:gaussian-random)
```
Comment (notmgsk) [inline src/matrix-operations.lisp]: us `a:` rather than `alexandria:`

File: tests/state-prep-tests.lisp [hunk 9]
```diff
@@ -53,19 +64,44 @@ CNOT 2 3
...
+
+(deftest test-state-prep-4q ()
+  (let* ((qubits (mapcar #'quil::qubit (list 0 1 2 3)))
+         (source-wf (quil::random-wavefunction 4))
+         (target-wf (quil::random-wavefunction 4))
+         (instr (make-instance 'quil::state-prep-application
+                               :arguments qubits
+                               :target-wf target-wf
+                               :source-wf source-wf)))
+    (let* ((output-matrix (quil::make-matrix-from-quil
```
Comment (appleby) [inline tests/state-prep-tests.lisp]: smush into previous `let*`?

File: src/compilers/state-prep.lisp [hunk 10]
```diff
@@ -232,6 +232,92 @@
...
+  (map '(vector (complex double-float))
+       (lambda (val) (coerce val '(complex double-float)))
+       elts))
+
+(defun schmidt-decomposition (phi num-a num-b)
+  "Given a wavefunction PHI containing subystems of size NUM-A and NUM-B qubits, compute the Schmidt decomposition of PHI.
+Returns c, U, V, where c is vector and U,V are unitary matrices (of dimensions
+NUM-A and NUM-B respectively) such that
+
+   PHI = \sum_{i} d_i U_i V _i
```
Comment (appleby) [inline src/compilers/state-prep.lisp]: c_i ?

File: src/compilers/state-prep.lisp (hunk 2 above)
Comment (kilimanjaro) [reply src/compilers/state-prep.lisp]: This object exists solely to feed to `MAGICL:SVD`, which takes a matrix.

File: tests/state-prep-tests.lisp (hunk 9 above)
Comment (kilimanjaro) [reply tests/state-prep-tests.lisp]: good call

File: src/compilers/state-prep.lisp (hunk 2 above)
Comment (notmgsk) [reply src/compilers/state-prep.lisp]: I mean a complex matrix (i.e. `make-complex-matrix`).

File: tests/state-prep-tests.lisp (hunk 6 above)
Comment (kilimanjaro) [reply tests/state-prep-tests.lisp]: `MAKE-COMPLEX-MATRIX` expects a list of entries. The main usage here is when WF is a complex vector, hence the call directly to `MAKE-MATRIX`. I'm updating the docstring to reflect this.

Comment (kilimanjaro) [reply tests/state-prep-tests.lisp]: (Most of the magicl api is pretty confusing re: lists, vectors, and matrices, but fixing that up is a different ballgame.)

File: src/compilers/state-prep.lisp [hunk 11]
```diff
@@ -232,6 +232,90 @@
...
+         (reshaped (magicl:transpose (magicl:make-matrix :rows size-a
+                                                         :cols size-b
+                                                         :data phi))))
+    (multiple-value-bind (u d vt)
+        (magicl:svd reshaped)
+      (values (coerce-to-complex-double-vector (magicl:matrix-diagonal d))
+              u
+              (magicl:transpose vt)))))
+
+(defun state-prep-4q (wf q0 q1 q2 q3 &key reversed)
```
Comment (ecpeterson) [inline src/compilers/state-prep.lisp]: I appreciate the inclusion of this!

**Comment by kilimanjaro** [comment]
> This is looking good.
> 
> Why did `test-state-prep-formation` break? The fix looks reasonable, but so does the original.

Actually I think it was only `test-state-prep-1q-source-and-target` that broke (since it didn't rescale by any prefactor). But since I was in there I updated other tests as well for the sake of consistency.

**Comment by ecpeterson** [comment]
Ohhh! Ok, excellent.

**Comment by kilimanjaro** [comment]
Tests pass locally. Aside from whatever stylistic improvements are called for, I think this is just waiting on the magicl v0.6.4 release.

**Comment by ecpeterson** [comment]
I released MAGICL; the changes between 0.6.3 and 0.6.4 are minor enough that I can take responsibility.

If this depends on the new MAGICL, be sure to bump the requirement in `cl-quil.asd`.

**Comment by notmgsk** [comment]
FYI I think this should not be merged until the new magicl version is available in quicklisp.

File: src/matrix-operations.lisp [hunk 12]
```diff
@@ -254,7 +269,7 @@ as needed so that they are the same size."
         (loop :for u :in ret :do
           (setf v (vector-difference v (vector-scale (dot-product v u) u))))
         (when (not (double= 0 (norm v)))
-          (setf ret (cons (vector-scale (/ 1 (norm v)) v) ret)))))
+          (setf ret (cons (vector-scale (/ (norm v)) v) ret)))))
```
Comment (appleby) [inline src/matrix-operations.lisp]: nit: `(push (vector-scale ...) ret)`

**Comment by kilimanjaro** [comment]
> FYI I think this should not be merged until the new magicl version is available in quicklisp.

For the sake of getting this closed I'm downgrading this to use the previous version of magicl. It still pre-emptively includes the fixed tests that magicl v0.6.4 would otherwise break, so later upgrading should be seamless.

File: src/compilers/state-prep.lisp [hunk 13]
```diff
@@ -232,6 +232,90 @@
...
                                      :arguments (list (first (application-arguments instr))))))))))
 
+(defun coerce-to-complex-double-vector (elts)
+  (map '(vector (complex double-float))
+       (lambda (val) (coerce val '(complex double-float)))
+       elts))
+
+(defun schmidt-decomposition (phi num-a num-b)
+  "Given a wavefunction PHI containing subystems of size NUM-A and NUM-B qubits, compute the Schmidt decomposition of PHI."
+  ;; Returns c, U, V, where c is vector and U, V are unitary matrices (of dimensions
```
Comment (notmgsk) [inline src/compilers/state-prep.lisp]: c is *a* vector

File: src/matrix-operations.lisp [hunk 14]
```diff
@@ -182,6 +182,21 @@ as needed so that they are the same size."
   (let ((m (magicl:random-unitary n)))
     (magicl:scale (expt (magicl:det m) (/ (- n))) m)))
 
+(defun random-wavefunction (n-qubits)
+  "Get a random complex unit vector with (EXPT 2 N-QUBITS) entries."
+  (let* ((size (expt 2 n-qubits)))
```
Comment (notmgsk) [inline src/matrix-operations.lisp]: unnecessary `let*`

File: src/compilers/state-prep.lisp (hunk 13 above)
Comment (kilimanjaro) [reply src/compilers/state-prep.lisp]: 🦅 👀

File: src/matrix-operations.lisp (hunk 14 above)
Comment (kilimanjaro) [reply src/matrix-operations.lisp]: I'm just flexing here

**Pull request merged**